### PR TITLE
178433179 implement xpowgate and ypowgate decomposition

### PR DESCRIPTION
**Pull request by dexter2206**
This PR implements gate decomposition engine. The engine, designed with @alexjuda , is inspired by formal grammars. Decomposition of gates (or more precisely, circuit operations) is defined via objects implementing `DecompositionRule`. Each such rule comprises:

- `predicate` determining whether this rule applies to given operation
- `production` converting given operation to iterable of one or more operations. Composing this sequence of operations should give an operation equivalent to the original one.

The engine provides generic `decompose_operations` function that can be used for constructing decomposer for (in principle) arbitrary third-party circuit library. Also, this PR provides decomposer for cirq circuits (`decompose_cirq_circuit`) and a single`PowerGateToPhaseAndRotation` decomposition rule. Depending on the configuration, this rule decomposes `YPowGate` and `XPowGate` instances that are not rotations or Pauli gates into a sequence:
```
PHASE(phi) * X * PHASE(phi) * X * ROTATION(theta)
```
where `theta` is the same as rotation angle in original gate, and `phi`, controlling global phase, is choosen such that the decomposition results in the same unitary as the original operation (see tests).

### CODE REVIEW

File: src/python/zquantum/core/circuits/_decomposition.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+from typing import Iterable, TypeVar
+
+from typing_extensions import Protocol
+
+OperationType = TypeVar("OperationType")
+
+
+class DecompositionRule(Protocol[OperationType]):
+    def production(self, operation: OperationType) -> Iterable[OperationType]:
+        raise NotImplementedError()
```
Comment (alexjuda) [inline src/python/zquantum/core/circuits/_decomposition.py:10]: WDYT about using "the `@abstractmethod` trick" so that codecov doesn't complain?

File: src/python/zquantum/core/circuits/_decomposition.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    def predicate(self, operation: OperationType) -> bool:
+        raise NotImplementedError()
+
+
+def decompose_operation(
+    operation: Iterable[OperationType],
+    decomposition_rules: Iterable[DecompositionRule[OperationType]],
+):
+    rules_iter = iter(decomposition_rules)
```
Comment (alexjuda) [inline src/python/zquantum/core/circuits/_decomposition.py:20]: I remember that forking iterators with [itertools.tee](https://docs.python.org/3/library/itertools.html#itertools.tee) is prone to performance problems if the forked iterators diverge much. I'm worried we might face a similar issue here when decomposing big circuits. WDYT?

File: src/python/zquantum/core/circuits/__init__.py [hunk 3]
```diff
@@ -141,7 +141,9 @@
     builtin_gate_by_name,
 )
 from ._circuit import Circuit
+from ._cirq_decompositions import PowerGateToPhaseAndRotation, decompose_cirq_circuit
 from ._compatibility import new_circuit_from_old_circuit
+from ._decomposition import DecompositionRule, decompose_operations
```
Comment (alexjuda) [inline src/python/zquantum/core/circuits/__init__.py:146]: The brand new decomposition engine seems to be unrelated to our Z-Quantum circuits framework. WDYT about moving it to a different place in the project, e.g. `zquantum.core.decomposition`?

**Review by alexjuda — CHANGES_REQUESTED** [review]

File: src/python/zquantum/core/circuits/_decomposition.py (hunk 1 above)
Comment (dexter2206) [reply src/python/zquantum/core/circuits/_decomposition.py:10]: Actually, it still does complain. I already tried this trick (see commit c535a50 and a18321f that reverts it). I think we would need to make the whole protocol an abstract class. Or, we could use `pass` instead of rising `NotImplementedError`. WDYT?

File: src/python/zquantum/core/circuits/_decomposition.py (hunk 2 above)
Comment (dexter2206) [reply src/python/zquantum/core/circuits/_decomposition.py]: Switch to using eager slices in 374c308

File: src/python/zquantum/core/circuits/_decomposition.py (hunk 1 above)
Comment (alexjuda) [reply src/python/zquantum/core/circuits/_decomposition.py]: The "trick" consists of marking methods with `@abstractmethod` decorator (it raises errors on its own) and removing their implementations, either replacing with `pass` or `...`. Would you try that?

Comment (dexter2206) [reply src/python/zquantum/core/circuits/_decomposition.py]: Hmm, I thought that the second part is unnecessary, but I'll try doing that!

File: src/python/zquantum/core/circuits/__init__.py (hunk 3 above)
Comment (dexter2206) [reply src/python/zquantum/core/circuits/__init__.py]: Moved to `zquantum.core.decomposition` in 36913a9

File: src/python/zquantum/core/decompositions/_decomposition.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def decompose_operation(
+    operation: Iterable[OperationType],
+    decomposition_rules: Sequence[DecompositionRule[OperationType]],
+):
+    if not decomposition_rules:
+        return [operation]
+
+    current_rule, remaining_rules = decomposition_rules[0], decomposition_rules[1:]
```
Comment (alexjuda) [inline src/python/zquantum/core/decompositions/_decomposition.py:31]: FYI, there's a destructuring syntax for this exact purpose :)
```suggestion
    current_rule, *remaining_rules = decomposition_rules
```

**Review by alexjuda — APPROVED** [review]

File: src/python/zquantum/core/decompositions/_decomposition.py (hunk 4 above)
Comment (dexter2206) [reply src/python/zquantum/core/decompositions/_decomposition.py]: I thought it is less known, but on the other hand it's much cleaner. Done in  e69e23e
